asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py

```diff
@@ -356,6 +356,114 @@ def create_router(self, context, router):
+        dynamic_nat_pool = None
+        if len(ext_ips) > 1:
+            dynamic_nat_pool = self._handle_dynamic_nat_pool_allocation(context, router_id, info, ext_ips)
+
+        # router = router or self._get_router(context, router_id)
+        result = super()._update_router_gw_info(context, router_id, info, request_body, router)
+        if dynamic_nat_pool is not None or orig_router_atts.dynamic_nat_pool:
+            self.db.set_router_dynamic_nat_pool(context, router_id, dynamic_nat_pool)
+
+        return result
+
+    def _handle_dynamic_nat_pool_allocation(self, context, router_id, info, ext_ips):
+        """Handle dynamic nat pooling, augment info with ip allocation, return nat pool
+
+        The caller will have to write back the NAT pool as router_att themselves
+        """
+        # someone is trying to add a gw with a nat pool (multiple ips)
+        if len(ext_ips) == 2:
+            raise asr1k_exc.DynamicNatPoolNeedsToHaveAtLeastTwoIPs()
+
+        has_ip_specified = False
+        has_no_ip_specified = False
+        alloc_subnet_id = None
+        for ip_def in ext_ips[:-1]:
+            # check if we have mixed ip/no-ip requests
+            if 'ip_address' in ip_def:
+                has_ip_specified = True
+            else:
+                has_no_ip_specified = True
+
+            # check if everything is the same subnet
+            if 'subnet_id' in ip_def:
+                if alloc_subnet_id is None:
+                    alloc_subnet_id = ip_def['subnet_id']
+                elif alloc_subnet_id != ip_def['subnet_id']:
+                    raise asr1k_exc.DynamicNatPoolTwoSubnetsFound(subnet_a=alloc_subnet_id,
+                                                                  subnet_b=ip_def['subnet_id'])
+        if has_ip_specified and has_no_ip_specified:
+            raise asr1k_exc.DynamicNatPoolIPDefinitionMixed()
+
+        dynamic_nat_pool = None
+        if has_no_ip_specified:
+            if not alloc_subnet_id:
+                raise asr1k_db.DynamicNatPoolIPDefinitionSubnetMissing()
+            ips, dynamic_nat_pool = self._find_ips_for_dynamic_nat_pool(context, alloc_subnet_id, len(ext_ips) - 1)
+            if not ips:
+                raise asr1k_exc.DynamicNatPoolExternalNetExhausted(ip_count=len(ext_ips) - 1, subnet_id=alloc_subnet_id)
+
+            for ext_ip, ip in zip(ext_ips, ips):
+                ext_ip['ip_address'] = ip
+            return dynamic_nat_pool
+        else:
+            # all ips are specified, check that they are consecutive
+            ips = [ip_def['ip_address'] for ip_def in ext_ips[:-1]]
+            ip = netaddr.IPAddress(ips[0])
+
+            for next_ip in ips[1:]:
+                next_ip = netaddr.IPAddress(next_ip)
+                if int(next_ip) - int(ip) != 1:
+                    raise asr1k_exc.DynamicNatPoolSpecifiedIpsNotConsecutivelyAscending(ip=ip, next_ip=next_ip)
+                ip = next_ip
```

Comment:
What about
```
In [17]: ips = [f'10.10.10.{i}' for i in range(10)]

In [18]: netaddr.IPSet(ips).iscontiguous()
Out[18]: True

In [19]: ips = [f'10.10.10.{i}' for i in range(10)] + ['10.10.10.20']

In [20]: netaddr.IPSet(ips).iscontiguous()
Out[20]: False

```

We could then also allow the user to provide them in whatever order (except the Interface IP), as long as they form a continuous space.

Reply:
So far I saw no reason to allow the user to give the IPs to us unsorted, I'd rather have them specify the pool in order, as I'd expect they do if they know what they're doing. Also, we would have to do something to detect addresses specified multiple times, as the set would deduplicate them. Or is that already done by neutron?

Reply:
I feel like neutron does no such thing in the `_update_gw_port()` methods. But you could easily compare `IPSet.size == len(ext_ip) - 1` to achieve that.

Reply:
Implemented as discussed